libdrgn/program.c

```diff
@@ -50,6 +50,27 @@ drgn_program_platform(struct drgn_program *prog)
 	return prog->has_platform ? &prog->platform : NULL;
 }
 
+void drgn_program_init_language(struct drgn_program *prog)
+{
+	struct drgn_error *err;
+	struct drgn_qualified_type main_type;
+	struct drgn_object res;
+
+	drgn_object_init(&res, prog);
+
+	// Use the CU from the "main" object as the lang for the whole program.
+	err = drgn_program_find_object(prog, "main", NULL,
+				       DRGN_FIND_OBJECT_ANY, &res);
+	if (!err) {
+		prog->lang = drgn_type_lang(res.type);
+		return;
+	}
+
+	// We could not deduce type or find result from "main".
+	// TODO have a better fallback here.
+	prog->lang = &drgn_language_c;
```

Comment:
Maybe we can just leave it unset if we couldn't find main, since `drgn_program_get_lang()` already has the concept of a fallback? Also, `drgn_object_deinit()` should be called here, and the error needs to be handled/destroyed. Something like

```c
err = drgn_program_find_object(...);
if (err) {
	if (err->code != DRGN_ERROR_LOOKUP)
		return err;
	/* We couldn't find "main". Don't set the language. */
	drgn_error_destroy(err);
} else {
	prog->lang = drgn_type_lang(res.type);
}
drgn_object_deinit(&res);
return NULL;
```